### DISCUSSION TITLE
What is the story with __slots__ support in Micropython

### DISCUSSION THREAD

**Discussion by brianreinhold** [micropython/micropython#13745] [General]
My understanding is that there has been some consideration to support the `__slots__` feature in Micropython but I don't know if that has happened, is in the works, or has been abandoned. Anyone know?

As a bonus, my understanding is that it is more efficient to use self variables in a class than passing parameters to instance methods BUT in defining self variables, does that take more memory?

**Comment by jimmo** [comment]
@brianreinhold There was a very old PR to prototype this, now abandoned, I haven't seen anybody looking into this recently.

It is potentially a good optimisation if you intend to create many instances of the same type. Essentially a given attribute on an instance takes up two words (for its key and value), and also you waste empty positions in the dictionary, whereas with `__slots__` you just have the keys for each instance. So a type with, say, six attributes could be (6+2)*2=16 words (64 bytes per instance) vs 6 words (rounded up to 8 for the GC block) (32 bytes per instance), so a net saving of 32 bytes per instance. (The attribute names are already interned, and obviously the values don't change).

Performance wise slots isn't a big gain, in fact it could be slower because you miss out on map caching.

I'm not quite sure what mean by the second part, maybe you could give two code examples to compare.

**Reply by brianreinhold** [reply]
Well, I will not try and use __slots__ then. Probably more important for a library.

What I meant by the second part was that I could create 'self' variables or I could pass the variables to methods as parameters, reducing the number of self variables. Right now I am doing a lot of the latter. That means creating more local stack space each method and cleaning up on return but no permanent space reserved for the instance variable. Will probably experiment.
